### PR TITLE
[PLGN-691] - GitHub - Full refresh of plugin to update SDK, add unit tests, update requirements and update dated code

### PR DESCRIPTION
**Pull request by rbowden-r7**
### Description

Describe the proposed changes:

 - Added connection test 
 - Updated to use the latest SDK version
 - Updated requirements.txt file
 - Updated to use Auth headers with git hub token rather than name and password in the auth section of the request 
 - Added unit tests
 - add better error handling 
 - updated the help.md file to add examples to all actions / trigger

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

new unit tests for all actions have been added 

```
----------------------------------------------------------------------
Ran 80 tests in 0.183s

OK
```

```
Name                                                                                                                        Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------------------------------------------------------------------------------------
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/__init__.py                                    0      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/__init__.py                           14      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/add_collaborator/__init__.py           1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/add_collaborator/action.py            27      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/add_collaborator/schema.py            19      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/add_issue_label/__init__.py            1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/add_issue_label/action.py             28      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/add_issue_label/schema.py             19      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/add_membership/__init__.py             1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/add_membership/action.py              27      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/add_membership/schema.py              23      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/block_user/__init__.py                 1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/block_user/action.py                  28      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/block_user/schema.py                  16      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/close_issue/__init__.py                1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/close_issue/action.py                 28      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/close_issue/schema.py                 18      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/create/__init__.py                     1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/create/action.py                      49      2    96%   85-86
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/create/schema.py                      22      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/create_issue_comment/__init__.py       1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/create_issue_comment/action.py        40      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/create_issue_comment/schema.py        19      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/get_issues_by_repo/__init__.py         1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/get_issues_by_repo/action.py          24      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/get_issues_by_repo/schema.py          17      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/get_my_issues/__init__.py              1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/get_my_issues/action.py               21      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/get_my_issues/schema.py               16      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/get_repo/__init__.py                   1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/get_repo/action.py                    24      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/get_repo/schema.py                    17      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/remove/__init__.py                     1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/remove/action.py                      51      1    98%   28
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/remove/schema.py                      18      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/search/__init__.py                     1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/search/action.py                      23      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/search/schema.py                      17      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/unblock_user/__init__.py               1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/unblock_user/action.py                26      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/unblock_user/schema.py                16      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/user/__init__.py                       1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/user/action.py                        18      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/actions/user/schema.py                        20      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/connection/__init__.py                         1      0   100%
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/connection/connection.py                      36     11    69%   33-34, 40-55
/Users/rbowden/insight_repos/insightconnect-plugins/plugins/github/komand_github/connection/schema.py                           8      0   100%
test_add_collaborator.py                                                                                                       34      0   100%
test_add_issue_label.py                                                                                                        34      0   100%
test_add_membership.py                                                                                                         34      0   100%
test_block_user.py                                                                                                             34      0   100%
test_close_issue.py                                                                                                            34      0   100%
test_create.py                                                                                                                 34      0   100%
test_create_issue_comment.py                                                                                                   34      0   100%
test_get_issues_by_repo.py                                                                                                     34      0   100%
test_get_my_issues.py                                                                                                          34      0   100%
test_get_repo.py                                                                                                               34      0   100%
test_remove.py                                                                                                                 34      0   100%
test_search.py                                                                                                                 34      0   100%
test_unblock_user.py                                                                                                           34      0   100%
test_user.py                                                                                                                   34      0   100%
util.py                                                                                                                       221     11    95%   167, 199, 226, 246, 281, 292-293, 298-301, 321
---------------------------------------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                                                        1441     25    98%
```

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

All actions and trigger have been tested using local docker container and postman

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``rapid7/insightconnect-python-3-38-slim-plugin:{sdk-version-num}`` and ``rapid7/insightconnect-python-3-38-plugin:{sdk-version-num}``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``insight-plugin validate`` which calls ``icon_validate`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `insight-plugin samples`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `insight-plugin validate` and make sure everything passes
2. Run the assessment tool: `insight-plugin run -A`. For single action validation: `insight-plugin run tests/{file}.json -A`
3. Copy (`insight-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/github/komand_github/actions/add_collaborator/action.py [hunk 1]
```diff
@@ -1,35 +1,67 @@
...
-            results = r.put(
-                "https://api.github.com/repos/{}/{}/collaborators/{}".format(
-                    params.get("organization"),
-                    params.get("repository"),
-                    params.get("username"),
-                ),
-                auth=self.connection.basic_auth,
-                params={"permission": params.get("permission")},
+            results = requests.put(
+                url=url, headers=self.connection.auth_header, params={"permission": permission}, timeout=60
```
Comment (ablakley-r7) [inline plugins/github/komand_github/actions/add_collaborator/action.py:36]: Can we default the timeout anywhere incase we have to modify in future?

File: plugins/github/komand_github/actions/add_collaborator/action.py [hunk 2]
```diff
@@ -1,35 +1,67 @@
...
             )
-            if results.status_code == 204:
-                return {"results": "User is already a collaborator"}
-            elif results.status_code == 201:
-                return {"results": results.json()}
-        except Exception as e:
-            self.logger.error("An error has occurred while adding collaborator: ", e)
-            raise
+            if results.status_code == 201:
```
Comment (ablakley-r7) [inline plugins/github/komand_github/actions/add_collaborator/action.py:38]: Similarly is there a way we could gather the status code handlers together or default some of their text? We might be able to use the values stored in the SDK if they are suitable.

File: plugins/github/komand_github/actions/search/action.py [hunk 3]
```diff
@@ -1,26 +1,47 @@
...
+            url = requests.compat.urljoin(self.connection.api_prefix, f"/search/{search_type}")
+            search_params = {"q": query}
+
+            results = requests.get(url=url, params=search_params, headers=self.connection.auth_header, timeout=60)
+
+            if results.status_code == 200:
+                return {Output.RESULTS: clean(results.json())}
+
+            else:
+                raise PluginException(
```
Comment (ablakley-r7) [inline plugins/github/komand_github/actions/search/action.py:34]: Can we also pass back data if it's suitable in the errors here which may help customers to debug their issue?

File: plugins/github/komand_github/connection/connection.py [hunk 4]
```diff
@@ -15,31 +18,38 @@ def __init__(self):
...
-            self.logger.error("Github: Connect: error %s", err.data)
-            raise Exception("Github: Connect: user could not be authenticated please try again.")
+            self.github_session = requests.Session()
 
-        except requests.exceptions.RequestException as e:
-            raise e
+        except github.GithubException:
+            raise PluginException(
+                cause="Github: Connect: user could not be authenticated please try again.",
+                assistance="Please check that your username and personal token is correct.",
```
Comment (ablakley-r7) [inline plugins/github/komand_github/connection/connection.py:36]: Can we pass back data here to help debug if suitable?

**Review by ablakley-r7 — COMMENTED** [review]
Looks great, basically a brand new plugin! Have a couple of comments on consolidating some pieces.

File: plugins/github/komand_github/actions/add_collaborator/action.py (hunk 1 above)
Comment (rbowden-r7) [reply plugins/github/komand_github/actions/add_collaborator/action.py]: I have added this as a common var to the util file

File: plugins/github/komand_github/actions/add_collaborator/action.py (hunk 2 above)
Comment (rbowden-r7) [reply plugins/github/komand_github/actions/add_collaborator/action.py]: I have added all generic error handling to the util file

File: plugins/github/komand_github/actions/add_collaborator/action.py (hunk 1 above)
Comment (rbowden-r7) [reply plugins/github/komand_github/actions/add_collaborator/action.py]: passing now passing back data in all cases

File: plugins/github/komand_github/actions/search/action.py (hunk 3 above)
Comment (rbowden-r7) [reply plugins/github/komand_github/actions/search/action.py]: I have added all generic error handling to the util file and passing back data

File: plugins/github/komand_github/connection/connection.py (hunk 4 above)
Comment (rbowden-r7) [reply plugins/github/komand_github/connection/connection.py:36]: I have added all generic error handling to the util file and passing back data

**Review by ablakley-r7 — APPROVED** [review]

File: plugins/github/komand_github/actions/close_issue/action.py [hunk 5]
```diff
@@ -14,19 +15,25 @@ def __init__(self):
...
+                repo = org.get_repo(repository)
+            else:
+                user = self.connection.user
+                repo = user.get_repo(repository)
+
+            issue = repo.get_issue(issue_number)
+            issue.edit(**issue_params)
+
+            return {Output.SUCCESS: True}
+
```
Comment (igorski-r7) [inline plugins/github/komand_github/actions/close_issue/action.py:37]:
```suggestion
            return {Output.SUCCESS: True}
```

File: plugins/github/komand_github/actions/create_issue_comment/action.py [hunk 6]
```diff
@@ -13,18 +22,26 @@ def __init__(self):
...
+
+            else:
+                user = self.connection.user
+                repo = user.get_repo(repository)
+
+            issue = repo.get_issue(issue_number)
+            issue_params = {"body": body}
+            issue = issue.create_comment(**issue_params)
+            return {"url": issue.html_url}
+
```
Comment (igorski-r7) [inline plugins/github/komand_github/actions/create_issue_comment/action.py:45]:
```suggestion
```

File: plugins/github/komand_github/actions/create_issue_comment/action.py [hunk 7]
```diff
@@ -13,18 +22,26 @@ def __init__(self):
...
+                repo = org.get_repo(repository)
+
+            else:
+                user = self.connection.user
+                repo = user.get_repo(repository)
+
+            issue = repo.get_issue(issue_number)
+            issue_params = {"body": body}
+            issue = issue.create_comment(**issue_params)
+            return {"url": issue.html_url}
```
Comment (igorski-r7) [inline plugins/github/komand_github/actions/create_issue_comment/action.py:44]:
```suggestion
            return {Output.URL: issue.html_url}
```

File: plugins/github/komand_github/actions/get_issues_by_repo/action.py [hunk 8]
```diff
@@ -1,26 +1,46 @@
...
+            title = urllib.parse.quote(params.get(Input.TITLE))
+            owner = urllib.parse.quote(params.get(Input.OWNER))
+
+            url = requests.compat.urljoin(self.connection.api_prefix, f"/repos/{owner}/{title}/issues")
+
+            results = requests.get(url=url, headers=self.connection.auth_header, timeout=TIMEOUT)
+
+            handle_http_exceptions(results)
+            return {Output.ISSUES: clean(results.json())}
+
```
Comment (igorski-r7) [inline plugins/github/komand_github/actions/get_issues_by_repo/action.py:37]:
```suggestion
```

File: plugins/github/komand_github/actions/get_my_issues/action.py [hunk 9]
```diff
@@ -1,21 +1,35 @@
...
-            results = requests.get("https://api.github.com/issues", auth=self.connection.basic_auth)
-            return {"issues": utils.clean(results.json())}
-        except Exception as e:
-            self.logger.error("Could not retrieve current user's assigned issues. Error: " + str(e))
+            url = requests.compat.urljoin(self.connection.api_prefix, "/issues")
+            results = requests.get(url=url, headers=self.connection.auth_header, timeout=TIMEOUT)
+
+            handle_http_exceptions(results)
+            return {"issues": clean(results.json())}
+
```
Comment (igorski-r7) [inline plugins/github/komand_github/actions/get_my_issues/action.py:26]:
```suggestion
            handle_http_exceptions(results)
            return {Output.ISSUES: clean(results.json())}
```

File: plugins/github/komand_github/actions/get_repo/action.py [hunk 10]
```diff
@@ -1,26 +1,40 @@
...
+            title = urllib.parse.quote(params.get(Input.TITLE))
+            owner = urllib.parse.quote(params.get(Input.OWNER))
+
+            url = requests.compat.urljoin(self.connection.api_prefix, f"/repos/{owner}/{title}")
+
+            results = requests.get(url=url, headers=self.connection.auth_header, timeout=TIMEOUT)
+
+            handle_http_exceptions(results)
+            return {Output.DATA: clean(results.json())}
+
```
Comment (igorski-r7) [inline plugins/github/komand_github/actions/get_repo/action.py:31]:
```suggestion
            handle_http_exceptions(results)
            return {Output.DATA: clean(results.json())}
```

File: plugins/github/komand_github/actions/unblock_user/action.py [hunk 11]
```diff
@@ -1,36 +1,42 @@
...
 
         try:
-            data = response.json()
-        except ValueError:
-            if response.status_code == 204:
-                self.logger.info("Successfully unblocked user")
-                return {"success": True}
+            response = requests.delete(url=url, headers=headers, timeout=TIMEOUT)
+            handle_http_exceptions(response)
+
```
Comment (igorski-r7) [inline plugins/github/komand_github/actions/unblock_user/action.py:31]:
```suggestion
```

File: plugins/github/komand_github/actions/user/action.py [hunk 12]
```diff
@@ -1,32 +1,41 @@
...
     def run(self, params={}):
-        g_user = github.Github()
-        user_info = g_user.get_user(params.get("username"))
-        user = {
-            "avatar": user_info.avatar_url,
-            "bio": user_info.bio,
-            "email": user_info.email,
-            "name": user_info.name,
-            "url": user_info.html_url,
+
```
Comment (igorski-r7) [inline plugins/github/komand_github/actions/user/action.py:16]:
```suggestion
```

**Review by igorski-r7 — APPROVED** [review]

**Review by jerichardson-r7 — APPROVED** [review]

**Pull request merged**